packages/opentelemetry-resources/src/Resource.ts

```diff
@@ -14,142 +14,127 @@
-  private _attributes?: Attributes;
+  private _rawAttributes: [string, MaybePromise<AttributeValue | undefined>][];
+  private _asyncAttributesPending = false;
 
-  /**
-   * Check if async attributes have resolved. This is useful to avoid awaiting
-   * waitForAsyncAttributes (which will introduce asynchronous behavior) when not necessary.
-   *
-   * @returns true if the resource "attributes" property is not yet settled to its final value
-   */
-  public asyncAttributesPending?: boolean;
-
-  /**
-   * Returns an empty Resource
-   */
-  static empty(): IResource {
-    return Resource.EMPTY;
-  }
+  private _memoizedAttributes?: Attributes;
 
+  public static EMPTY = new Resource({ attributes: {} });
   /**
    * Returns a Resource that identifies the SDK in use.
    */
   static default(): IResource {
     return new Resource({
-      [SEMRESATTRS_SERVICE_NAME]: defaultServiceName(),
-      [SEMRESATTRS_TELEMETRY_SDK_LANGUAGE]:
-        SDK_INFO[SEMRESATTRS_TELEMETRY_SDK_LANGUAGE],
-      [SEMRESATTRS_TELEMETRY_SDK_NAME]:
-        SDK_INFO[SEMRESATTRS_TELEMETRY_SDK_NAME],
-      [SEMRESATTRS_TELEMETRY_SDK_VERSION]:
-        SDK_INFO[SEMRESATTRS_TELEMETRY_SDK_VERSION],
+      attributes: {
+        [ATTR_SERVICE_NAME]: defaultServiceName(),
+        [ATTR_TELEMETRY_SDK_LANGUAGE]: SDK_INFO[ATTR_TELEMETRY_SDK_LANGUAGE],
+        [ATTR_TELEMETRY_SDK_NAME]: SDK_INFO[ATTR_TELEMETRY_SDK_NAME],
+        [ATTR_TELEMETRY_SDK_VERSION]: SDK_INFO[ATTR_TELEMETRY_SDK_VERSION],
+      },
     });
   }
 
+  static FromAttributeList(
+    attributes: [string, MaybePromise<AttributeValue | undefined>][]
+  ): Resource {
+    const res = new Resource({});
+    res._rawAttributes = attributes;
+    res._asyncAttributesPending =
+      attributes.filter(([_, val]) => isPromiseLike(val)).length > 0;
+    return res;
+  }
+
   constructor(
     /**
      * A dictionary of attributes with string keys and values that provide
      * information about the entity as numbers, strings or booleans
      * TODO: Consider to add check/validation on attributes.
      */
-    attributes: Attributes,
-    asyncAttributesPromise?: Promise<Attributes>
+    resource: DetectedResource
   ) {
-    this._attributes = attributes;
-    this.asyncAttributesPending = asyncAttributesPromise != null;
-    this._syncAttributes = this._attributes ?? {};
-    this._asyncAttributesPromise = asyncAttributesPromise?.then(
-      asyncAttributes => {
-        this._attributes = Object.assign({}, this._attributes, asyncAttributes);
-        this.asyncAttributesPending = false;
-        return asyncAttributes;
-      },
-      err => {
+    const attributes = resource.attributes ?? {};
+    this._rawAttributes = Object.entries(attributes).map(([k, v]) => {
+      if (isPromiseLike(v)) {
+        // side-effect
+        this._asyncAttributesPending = true;
+      }
+
+      return [k, v];
+    });
+  }
+
+  public get asyncAttributesPending() {
```

Comment:
:+1: on making this a getter – should this field on `IResource` be marked as `readonly`? I suppose that was always the intent, but it isn't currently